background.js

```diff
@@ -1,31 +1,66 @@
 "use strict";
 
-/* global chrome, URL */
+/* global chrome */
 
 chrome.runtime.onInstalled.addListener(function () {
-  chrome.storage.local.get(["blocked", "enabled"], function (local) {
-    if (!Array.isArray(local.blocked)) {
-      chrome.storage.local.set({ blocked: [] });
-    }
-
+  chrome.storage.local.get(["enabled", "blocked"], function (local) {
     if (typeof local.enabled !== "boolean") {
       chrome.storage.local.set({ enabled: false });
     }
+
+    if (!Array.isArray(local.blocked)) {
+      chrome.storage.local.set({ blocked: [] });
+    }
   });
 });
 
+// "https://www.youtube.com/" => "youtube.com"
+// "https://www.youtube.com/feed/explore" => "youtube.com/feed/explore"
+// "https://music.youtube.com/" => "music.youtube.com"
+// "https://music.youtube.com/explore" => "music.youtube.com/explore"
+const getDomainUrl = (url) => {
+  let domainUrl = url.replace(/^http(s?):\/\//, "").replace(/^www\./, "");
+  domainUrl = domainUrl.endsWith("/") ? domainUrl.slice(0, -1) : domainUrl;
+  return domainUrl;
+};
+
+// ["youtube.com", "!music.youtube.com"] => [{ path: "youtube.com", type: "block" }, { path: "music.youtube.com", type: "allow" }]
+const getRules = (blocked) => {
+  const allowList = blocked
+    .filter((item) => item.startsWith("!"))
+    .map((item) => item.substring(1));
+
+  const blockList = blocked
+    .filter((item) => !item.startsWith("!"));
+
+  const rules = [
+    ...allowList.map((path) => ({ path, type: "allow" })),
+    ...blockList.map((path) => ({ path, type: "block" })),
+  ].sort((a, b) => b.path.length - a.path.length);
```

Comment:
I am sorting the rules by their length, as their length defines their specificity.
**Example:** `music.youtube.com` is more specific than `youtube.com` and we might want to allow `music.youtube.com` while still block `youtube.com`, so we should check `music.youtube.com` rule first (otherwise `youtube.com` rule would block `music.youtube.com`).

Reply:
Sorting the rules by length is an elegant solution to order by specificity!

Reply:
Thanks!